accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,309 @@
+- Optional information regarding the health of each node of the cluster (if supported by the cluster).
+
+By enumerating all the functions of the Salt module corresponding to the cluster provider manager, Uyuni/SUSE Manager will offer to run all the actions that the cluster provider manager exposes.
+A minimum list of actions that must be exposed by the cluster provider manager is:
+
+* Adding a node: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) (e.g. username, IP of the system, type of the node).
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.add_node <cluster name> <optional params>
+  ```
+
+  All the parameters of the formula have to be specified by the cluster provider manager's types.
+
+  In a future version, Uyuni/SUSE Manager may offer the possibility to present to the user a drop-down selection of systems registered to Uyuni/SUSE Manager and not part of any cluster.
+
+* Removing a node: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.remove_node <cluster name> <optional params>
+  ```
+
+* Node reboot: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.reboot_node <cluster name> <optional params>
+  ```
+
+  The implementation of how a node is rebooted is left to the specific cluster provider manager.
+
+* Cluster upgrade: As every cluster has its peculiarities when upgrading the cluster nodes and Uyuni/SUSE Manager should be agnostic about the implementation strategy, Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.upgrade_cluster <cluster name> <optional params>
+  ```
+
+### Single cluster node management with Uyuni/SUSE Manager
+
+The following actions refer to the individual cluster node registered to Uyuni/SUSE Manager. The following actions must be implemented after the cluster overview and its corresponding actions are implemented.
+NOTE: registering an SES 6 node to Uyuni/SUSE Manager is not be possible as SES 6 management node is a Salt master. From SES 7 on (under development), Salt will be used for deployment only.
+
+#### Relaxing system lock
+
+In the cluster overview, when listing individual nodes of the cluster, Uyuni/SUSE Manager will also offer the possibility of registering an individual node to Uyuni/SUSE Manager (with System > Bootstrapping).
```

Comment:
I have a use case for freeing up a node (and thus making it generic again). I don't see one for just unregistering a single node from SUSE Manager.

Reply:
> Will it be possible to delete a node from a cluster (=unregister it from Uyuni but still have it in the cluster), or detach a node from a cluster (=keep it in Uyuni but remove it from the cluster)?

As we imagined and hopefully conveyed in the RFC:
- When we remove a node from the cluster, we are not deleting ("unregistering") the single node from Uyuni/SUSE Manager
- When we delete ("unregister") a single node from Uyuni/SUSE Manager, we are not removing the node from the cluster

Reply:
> > Will it be possible to delete a node from a cluster (=unregister it from Uyuni but still have it in the cluster), or detach a node from a cluster (=keep it in Uyuni but remove it from the cluster)?
> 
> As we imagined and hopefully conveyed in the RFC:
> 
>     * When we remove a node from the cluster, we are not deleting ("unregistering") the single node from Uyuni/SUSE Manager

OK, freeing up a node makes sense. E. g. I may take a CaaSP node, free it from the CaaSP cluster, then convert it to an HA node.

>     * When we delete ("unregister") a single node from Uyuni/SUSE Manager, we are not removing the node from the cluster

Does it make sense to have a cluster with e. g. 5 nodes and then only show 4 of them in Uyuni because one the nodes has been unregistered from Uyuni? What's the use case? I think Uyuni should show all the nodes in the cluster.

Reply:
> > ```
> > * When we delete ("unregister") a single node from Uyuni/SUSE Manager, we are not removing the node from the cluster
> > ```
> 
> 
> Does it make sense to have a cluster with e. g. 5 nodes and then only show 4 of them in Uyuni because one the nodes has been unregistered from Uyuni? What's the use case? I think Uyuni should show all the nodes in the cluster.

If you unregister a single node from Uyuni/SUSE Manager, the node will still be listed in the Cluster Overview. You will not see the single node in the system list, the list of systems registered to Uyuni/SUSE Manager.
Am I missing something?

Reply:
IMHO, unregistering without also taking the node out of the logical cluster makes no sense. So we shouldn't allow/encourage it. But for the MVP we don't have to implement an "auto-cleanup" as in "node was removed in SUSE Manager, so we have to trigger node removal via the cluster API". It's ok to document that removing a node via the API keeps it in SUSE Manager (albeit in a neutral system group), while removing it just in SUSE Manager will have no immediate effect on the cluster.

Reply:
OK

Reply:
Added the two clarification points as discussed.